### PR TITLE
vtexplain groundwork

### PR DESCRIPTION
**Pull request by demmer**
Minor changes needed to support vtexplain:

* Add an optional QueryLogger to the fakesqldb that calls a registered callback whenever 
* Add an accessor to expose the vtgate executor's plan cache outside of the package
* Expose a constructor to create a vttablet QueryExecutor and to get the tablet execution plan

### CODE REVIEW

**Comment by demmer** [comment]
@sougou It's been challenging to get Travis to pass on this PR, but I also have a very hard time seeing how any of these changes could have actually introduced a regression since all they do is add hooks for future use by vtexplain. Can you take a look?

File: go/vt/vtgate/executor.go [hunk 1]
```diff
@@ -747,6 +747,13 @@ func (e *Executor) ServeHTTP(response http.ResponseWriter, request *http.Request
 	}
 }
 
+// Plans returns the LRU plan cache
+func (e *Executor) Plans() *cache.LRUCache {
+	e.mu.Lock()
```
Comment (sougou) [inline go/vt/vtgate/executor.go]: No need to lock. `e.plans` has built-in synchronization.

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 2]
```diff
@@ -58,6 +58,20 @@ type QueryExecutor struct {
 	tsv              *TabletServer
 }
 
+// NewQueryExecutor creates a new executor.
+func NewQueryExecutor(ctx context.Context, query string, bindVars map[string]*querypb.BindVariable, transactionID int64, options *querypb.ExecuteOptions, plan *TabletPlan, logStats *tabletenv.LogStats, tsv *TabletServer) *QueryExecutor {
```
Comment (sougou) [inline go/vt/vttablet/tabletserver/query_executor.go]: You intend to change the call sites for this later?

File: go/mysql/fakesqldb/server.go [hunk 3]
```diff
@@ -290,6 +293,20 @@ func (db *DB) ConnectionClosed(c *mysql.Conn) {
 
 // ComQuery is part of the mysql.Handler interface.
 func (db *DB) ComQuery(c *mysql.Conn, q []byte, callback func(*sqltypes.Result) error) error {
+	err := db.comQueryImpl(c, q, func(qr *sqltypes.Result) error {
```
Comment (sougou) [inline go/mysql/fakesqldb/server.go]: One problem you'll have to think about is results that lead to other results. Specifically, how to make lookup vindexes return meaningful values so the executor follows through on the routing.

**Comment by sougou** [comment]
Looks like your build went green.

**Comment by demmer** [comment]
I decided in hindsight that having two separate PRs for the vtexplain groundwork and the actual implementation was hindering as opposed to helping the review. I'm going to close this one as I've moved these commits into #3082.

**Pull request closed without merging**